### PR TITLE
add isIdentified property and check on identify()

### PR DESCRIPTION
**Pull request by beradeep**
## :bulb: Motivation and Context

This change adds the is_identified property (absent earlier) to check user is identified or not (anonymous). This change also adds a fix to identify() to ensure that an "identify" event is captured if and only if the user isn't already identified (anonymous). Also adds an in-memory flag to check if the property is already loaded or not. 

fix https://github.com/PostHog/posthog-android/issues/136 & https://github.com/PostHog/posthog-android/issues/108

## :green_heart: How did you test it?
- Added unit test to assert against duplicate "identify" captures
- Added unit test to assert "identify" works as usual on reset
- Verified that this change doesn't break existing tests

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

### CODE REVIEW

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 1]
```diff
@@ -227,6 +231,24 @@ public class PostHog private constructor(
             getPreferences().setValue(DISTINCT_ID, value)
         }
 
+    private var isIdentified: Boolean = false
+        @JvmName("get-isIdentified")
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:235]: we don't need this

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 2]
```diff
@@ -227,6 +231,24 @@ public class PostHog private constructor(
...
+        get() {
+            synchronized(identifiedLock) {
+                if (!isIdentifiedLoaded) {
+                    println("hi")
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:239]: we can remove this

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 3]
```diff
@@ -227,6 +231,24 @@ public class PostHog private constructor(
...
+                    this.isIdentified = getPreferences().getValue(IS_IDENTIFIED) as? Boolean
+                        ?: (this.distinctId != this.anonymousId)
+                    isIdentifiedLoaded = true
+                }
+            }
+            return field
+        }
+        set(value) {
+            field = value
+            getPreferences().setValue(IS_IDENTIFIED, value)
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:249]: I guess we should sync this with `identifiedLock` as well

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 4]
```diff
@@ -484,27 +506,30 @@ public class PostHog private constructor(
...
             }
             this.distinctId = distinctId
+            this.isIdentified = true
 
             // only because of testing in isolation, this flag is always enabled
             if (reloadFeatureFlags) {
                 reloadFeatureFlags()
             }
+        } else {
+            config?.logger?.log("already identified with id : $distinctId.")
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:532]:
```suggestion
            config?.logger?.log("already identified with id: $distinctId.")
```

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 5]
```diff
@@ -707,6 +732,7 @@ public class PostHog private constructor(
             return
         }
 
+        isIdentifiedLoaded = false
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:735]: we should sync this with `identifiedLock`

**Comment by marandaneto** [comment]
Thanks @beradeep 
I left a few comments, after addressing, we can merge it
don't forget to add an entry to the changelog.

File: posthog/src/main/java/com/posthog/PostHog.kt (hunk 2 above)
Comment (beradeep) [reply posthog/src/main/java/com/posthog/PostHog.kt:239]: Oh, my bad, I'm sorry I forgot to omit that.

File: posthog/src/main/java/com/posthog/PostHog.kt (hunk 3 above)
Comment (beradeep) [reply posthog/src/main/java/com/posthog/PostHog.kt:249]: okay 👍

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 6]
```diff
@@ -707,6 +732,9 @@ public class PostHog private constructor(
             return
         }
 
+        synchronized(identifiedLock) {
+            isIdentifiedLoaded = false
+        }
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:737]: This should go in the reset method and not here

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 7]
```diff
@@ -484,27 +506,30 @@ public class PostHog private constructor(
...
         }
 
-        capture(
-            "\$identify",
-            distinctId = distinctId,
-            properties = props,
-            userProperties = userProperties,
-            userPropertiesSetOnce = userPropertiesSetOnce,
-        )
+        if (previousDistinctId != distinctId && !this.isIdentified) {
```
Comment (marandaneto) [inline posthog/src/main/java/com/posthog/PostHog.kt:509]: I think the behavior here is different, see the JS SDK https://github.com/PostHog/posthog-js/blob/e8a645134a03477d29ab6839ae7f10fcd6e4157e/src/posthog-core.ts#L1316-L1354
if it's the very same `distinctId` again won't capture anything (and capture a $set event if needed)
if it's a new distinctId, clear all the cache from the previous distinctId, and identify again with the new distinctId, reload feature flags, etc

Comment (beradeep) [reply posthog/src/main/java/com/posthog/PostHog.kt:509]: Isn't it nearly the same behavior, except for the `setPersonProperties` and `setPersonPropertiesForFlags`, which are not yet implemented here.

Comment (marandaneto) [reply posthog/src/main/java/com/posthog/PostHog.kt:509]: nvm, I think I misread the PR, you are right.

**Comment by marandaneto** [comment]
The flag `$is_identified` isn't being set though, so this issue can't be closed https://github.com/PostHog/posthog-android/issues/136

File: posthog/src/main/java/com/posthog/PostHog.kt (hunk 6 above)
Comment (marandaneto) [reply posthog/src/main/java/com/posthog/PostHog.kt]: i pushed 2 commits to fix this

**Comment by marandaneto** [comment]
> The flag `$is_identified` isn't being set though, so this issue can't be closed #136

i pushed 2 commits to fix this

**Review by marandaneto — APPROVED** [review]

**Pull request merged**